### ISSUE TITLE
Error in the logging when for the protocol version no template exists

### ISSUE THREAD

**Issue by klvries** [OSGP/Platform#162]
Handle shipment file: nullpointer exception in the logging when for the protocol version no template exitsts.

I created a new version 5.0.1 in the table osgp_core.protocol_info.

When I use this protocol version in the shipment file this results in a nullpointer exception logging for the sending of the notification. It should be better that in the logging was mentioned that no template exists for the protocol version.

Now the following logging is shown:
```
11:04:25.092 [DefaultMessageListenerContainer-3] WARN  (AbstractMessageListenerContainer.java:936) o.s.j.l.DefaultMessageListenerContainer - Execution of JMS message listener failed, and no ErrorHandler has been set.
java.lang.NullPointerException: null
        at com.alliander.osgp.core.infra.jms.protocol.ProtocolRequestMessageSender.sendMessage(ProtocolRequestMessageSender.java:57) ~[ProtocolRequestMessageSender.class:na]
        at com.alliander.osgp.core.infra.jms.protocol.ProtocolRequestMessageSender.send(ProtocolRequestMessageSender.java:51) ~[ProtocolRequestMessageSender.class:na]
        at com.alliander.osgp.core.application.services.DeviceRequestMessageService.processMessage(DeviceRequestMessageService.java:80) ~[DeviceRequestMessageService.class:na]
        at com.alliander.osgp.core.application.services.DeviceRequestMessageService$$FastClassBySpringCGLIB$$5aec3c42.invoke(<generated>) ~[spring-core-4.2.0.RELEASE.jar:na]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:717) ~[spring-aop-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99) ~[spring-tx-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281) ~[spring-tx-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96) ~[spring-tx-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:653) ~[spring-aop-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at com.alliander.osgp.core.application.services.DeviceRequestMessageService$$EnhancerBySpringCGLIB$$a17c0e02.processMessage(<generated>) ~[spring-core-4.2.0.RELEASE.jar:na]
        at com.alliander.osgp.core.infra.jms.domain.DomainRequestMessageListener.onMessage(DomainRequestMessageListener.java:58) ~[DomainRequestMessageListener.class:na]
        at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:746) ~[spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:684) ~[spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:651) ~[spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:315) [spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:253) [spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1150) [spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1142) [spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1039) [spring-jms-4.2.0.RELEASE.jar:4.2.0.RELEASE]
        at java.lang.Thread.run(Thread.java:744) [na:1.7.0_55]
11:05:00.000 [osgp-core-scheduling-1] INFO  (ScheduledTaskScheduler.java:43) c.a.o.c.a.t.ScheduledTaskScheduler - Processing scheduled tasks
11:05:04.166 [pool-10-thread-1] INFO  (ShipmentFileHandler.java:245) c.a.o.s.w.i.io.ShipmentFileHandler - Finished processing /home/flexadmin/shipment/unencrypted/done/unencrypted_20150930-11:03:52.xml
11:05:04.166 [pool-10-thread-1] INFO  (ShipmentFileCleanupThread.java:35) c.a.o.s.w.i.i.ShipmentFileCleanupThread - Delay over for thread of shipmentFile: 54
```

The device records are all witten to the DB. 
The xml is moved to the done direcry with the message in progrees for the devices.

**Comment by thijlboonen** [comment]
I can't reproduce this bug.

I added a DSMR 5.0.1 to the osgp_core.protocol_info database.

In the shipmentfile, I replaced the version with `<tns:DSMR_version>5.0.1</tns:DSMR_version>` 

The device is added correctly, no nullpointer exception was thrown.

**Comment by kevinsmeets** [comment]
@thijlboonen Let me try to help you out.


Long version:

Let's say somebody adds protocol-info record for the other protocol-version to the  protocol-info table. Then, without restarting tomcat, he tries to add a device using this other protocol

When a request is issued for this device, the osgp-core project will try to find a protocol info object for the device, which can be present too (especially if you added it yourself...).

Now, when this protocol info object contains a protocol-version which was not present when tomcat started, the **JmsTemplate** needed to send a queue message to the requests-queue of a protocol-adapter will be **null**. (Because you need to restart tomcat in order to reload the ApplicationContext before the ProtocolRequestMessageJmsTemplateFactory will contain the JmsTemplate for the requests-queue for the other protocol. It will not just appear out of thin air! :) )


Short version:

look at the line of code mentioned in the stack-trace
java.lang.NullPointerException: null
        at com.alliander.osgp.core.infra.jms.protocol.ProtocolRequestMessageSender.sendMessage(ProtocolRequestMessageSender.java:57)

The call of a function of a JmsTempate (which is null) at line 57.


Fix:

Perhaps a null check in ProtocolRequestMessageJmsTemplateFactory.getTemplate() ?
Or in ProtocolRequestMessage.send() would work too!

**Comment by klvries** [comment]
The issue is still there when you use a not existing device_identification

**Comment by rlemmers** [comment]
This issue only occurs when someone manually updates the database, which makes the priority of this issue very low. I will move the item to the backlog.
